Lib/subprocess.py

```diff
@@ -1414,6 +1419,16 @@ def _execute_child(self, args, executable, preexec_fn, close_fds,
 
             if executable is None:
                 executable = args[0]
+
+            if (preexec_fn is None
+                    and not close_fds
+                    and not pass_fds
+                    and cwd is None
+                    and env is None
+                    and restore_signals is False
+                    and not start_new_session):
```

Comment:
Right. I suggest to add more tests:

```
and os.path.isabs(executable)
and self.stdin is self.stdout is self.stderr is None)
```

Reply:
> Looks like some tests that would catch the above are needed :)

test_subprocess is quite complete. But with this PR, posix_spawn() is only used under very specific conditions: restore_signals=False and close_fds=False. This case is rare.

@nanjekyejoannah is my mentoree and we are working together to slowly use posix_spawn() in more and more cases. I prefer to work step by step, and so support more and more options, but start from the bare minimum, and check at each step that everything is fine.

Reply:
`os.path.isabs(executable)` is a bit strong. Using `os.sep not in executable` is enough: it's exactly what `execvp` and `posix_spawnp` do when they decide whether to do `PATH` lookup. But IMO even better would be to add `os.posix_spawnp` and use it instead.

> test_subprocess is quite complete. But with this PR, posix_spawn() is only used under very specific conditions: restore_signals=False and close_fds=False. This case is rare.

It's a bit scary for me that basic functionality broke (albeit under these rare conditions) and no test failed. Basically, it indicates that if users pass `restore_signals=False`, they're on their own (`close_fds` seems to be better tested).

Reply:
Don't worry, os.posix_spawn() support will be much more complete before Python 3.8 final. I'm working with @nanjekyejoannah to support restore_signals=True for example ;-)

Reply:
> But IMO even better would be to add os.posix_spawnp and use it instead.

Aha, I didn't know posix_spawnp(), that's interesting. It is available on macOS and FreeBSD.

Again, I prefer to keep the first implementation as simple as possible. os.posix_spawnp() can be added later.

> os.path.isabs(executable) is a bit strong. Using os.sep not in executable is enough: it's exactly what execvp and posix_spawnp do when they decide whether to do PATH lookup.

shutil.which() uses os.path.dirname(). So I suggest to check "and os.path.dirname(executable)" instead of "and os.path.isabs(executable)".

Reply:
Update, according to the discussion, the following check would be better:

```
and os.path.dirname(executable)
and self.stdin is self.stdout is self.stderr is None)
```